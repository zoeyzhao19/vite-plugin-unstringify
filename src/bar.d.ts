declare function func(name: string): void;
declare type A = string;
