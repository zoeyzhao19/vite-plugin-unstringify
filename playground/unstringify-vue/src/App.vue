<script setup lang="ts">
import TestTsx from './test';
import Another from './another.vue'
</script>

<template>
  <div :data-aaac="{
    type: 'type_a',
    name: 'name_a',
    desc: 'desc_a',
  }">
    <p :data-bbb="{
      type: 'type_b',
      name: 'name_b',
      desc: 'desc_b',
    }">
      <TestTsx></TestTsx>
    </p>
  </div>
</template>

<style scoped>
.logo {
  height: 6em;
  padding: 1.5em;
  will-change: filter;
}

.logo:hover {
  filter: drop-shadow(0 0 2em #646cffaa);
}

.logo.vue:hover {
  filter: drop-shadow(0 0 2em #42b883aa);
}
</style>
