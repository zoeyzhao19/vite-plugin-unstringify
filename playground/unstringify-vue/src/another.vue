<template>
  <div :data-aaa="{
    type: 'type_a',
    name: 'name_a',
    desc: 'desc_a',
  }">
    <p :data-bbb="{
      type: 'type_b',
      name: 'name_b',
      desc: 'desc_b',
    }">
      another
    </p>
  </div>
</template>