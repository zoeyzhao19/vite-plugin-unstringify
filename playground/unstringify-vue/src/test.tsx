export default function TestTsx() {
  return (
    <div
      data-jt-sender={{
        type: 'clickAndSpm',
        event_type: 'click',
        block_desc: 'articleLibrarySearch',
        button_desc: 'enter_btn',
      }}
    >
      323232
    </div>
  );
}
