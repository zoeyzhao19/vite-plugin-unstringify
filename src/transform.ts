import MagicString from 'magic-string';

export function transformVueKey(
  code: string,
  vueKeyREMap: Map<string | number, RegExp>
) {
  const s = new MagicString(code);
  for (const [, vueKeyRE] of vueKeyREMap) {
    while (true) {
      const matches = vueKeyRE.exec(code);
      if (!matches) break;
      if (matches && matches[1] && matches[2]) {
        s.overwrite(
          matches['index'] + matches[1].length + 3,
          matches['index'] + matches[1].length + 3 + matches[2].length + 2,
          `JSON.stringify({${matches[2]}})`
        );
      }
    }
    vueKeyRE.lastIndex = 0;
  }
  if (s.hasChanged()) {
    return {
      code: s.toString(),
      map: s.generateMap(),
    };
  }
  return null;
}

export function transformJsxKey(
  code: string,
  jsxKeyREMap: Map<string | number, RegExp>
) {
  const s = new MagicString(code);
  for (const [, jsxKeyRE] of jsxKeyREMap) {
    while (true) {
      const matches = jsxKeyRE.exec(code);
      if (!matches) break;
      if (matches && matches[1] && matches[2]) {
        s.overwrite(
          matches['index'] + matches[1].length + 2,
          matches['index'] + matches[1].length + 2 + matches[2].length + 2,
          `JSON.stringify({${matches[2]}})`
        );
      }
    }
    jsxKeyRE.lastIndex = 0;
  }
  if (s.hasChanged()) {
    return {
      code: s.toString(),
      map: s.generateMap(),
    };
  }
  return null;
}
