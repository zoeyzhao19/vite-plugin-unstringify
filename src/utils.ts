export function escapeRegExp(str: string) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export function resolveDataKey(rawStr: string, isEntry = true): string {
  if (isEntry) {
    rawStr = rawStr.startsWith('^') ? rawStr.slice(1) : rawStr;
    rawStr = rawStr.endsWith('$') ? rawStr.slice(0, -1) : rawStr;
  }
  const greedyRE = /[^\\][?]$/;
  if (greedyRE.test(rawStr)) {
    return resolveDataKey(rawStr.slice(0, -1), false);
  }
  return rawStr;
}

export const generateKeyFromRe = (rawStr: string) => {
  const greedyRE = /[^\\][*+]$/;
  if (greedyRE.test(rawStr)) {
    return rawStr.slice(0, -1);
  }
};

export function vueKeyRe(rawStr: string) {
  return new RegExp(`:(${rawStr})=["']\\{([\\s\\S]+?)}[\\s]*["']`, 'g');
}

export function jsxKeyRe(rawStr: string) {
  return new RegExp(`(${rawStr})={{([\\s\\S]+?)}}`, 'g');
}
