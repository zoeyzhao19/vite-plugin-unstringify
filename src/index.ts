import { resolveDataKey, vueKeyRe, jsxKeyRe } from './utils';
import { transformVueKey, transformJsxKey } from './transform';

const vueKeyREMap = new Map<string | number, RegExp>();
const jsxKeyREMap = new Map<string | number, RegExp>();

export default function unstringify(dataKey: RegExp | string | string[]): {
  name: string;
  enforce?: 'pre' | 'post';
  transform(
    code: string,
    id: string
  ): { code: string; map: any } | null | string;
} {
  if (Object.prototype.toString.call(dataKey) === '[object RegExp]') {
    const dataKeyStr = resolveDataKey(dataKey.toString().slice(1, -1));
    vueKeyREMap.set(0, vueKeyRe(`${dataKeyStr}`));
    jsxKeyREMap.set(0, jsxKeyRe(`${dataKeyStr}`));
  } else {
    const tempDateKeys = Array.isArray(dataKey) ? dataKey : [dataKey];
    tempDateKeys.forEach((key, index) => {
      vueKeyREMap.set(index, vueKeyRe(key as string));
      jsxKeyREMap.set(index, jsxKeyRe(key as string));
    });
  }

  return {
    name: 'vite-plugin-unstringify',
    enforce: 'pre',
    transform(code: string, id: string) {
      let result;
      if (id.endsWith('vue')) {
        result = transformVueKey(code, vueKeyREMap);
      } else if (id.endsWith('jsx') || id.endsWith('tsx')) {
        result = transformJsxKey(code, jsxKeyREMap);
      }

      return result ?? null;
    },
  };
}
