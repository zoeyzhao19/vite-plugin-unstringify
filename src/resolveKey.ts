import { isRegExp } from './utils';

export function resolveDataKey(dataKey: RegExp | string | string[]) {
  let resolveKeys: string[] = [];
  function resolveRE(rawStr: string, isEntry = true): string {
    if (isEntry) {
      rawStr = rawStr.startsWith('^') ? rawStr.slice(1) : rawStr;
      rawStr = rawStr.endsWith('$') ? rawStr.slice(0, -1) : rawStr;
    }
    const greedyRE = /[^\\][?]$/;
    if (greedyRE.test(rawStr)) {
      return resolveRE(rawStr.slice(0, -1), false);
    }
    return rawStr;
  }

  if (isRegExp(dataKey)) {
    resolveKeys = [resolveRE(dataKey.toString().slice(1, -1))];
  } else {
    resolveKeys = Array.isArray(dataKey) ? dataKey : [dataKey];
  }

  return resolveKeys;
}
