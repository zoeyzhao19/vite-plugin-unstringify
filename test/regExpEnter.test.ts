import { describe, test, expect } from 'vitest';
import { vueKeyRe, jsxKeyRe } from '../src/utils';
import { transformVueKey, transformJsxKey } from '../src/transform';
import { resolveDataKey } from '../src/resolveKey';

describe('transformVueKey', () => {
  const code = `<div :data-aaa="{
    type: 'type_a',
    name: 'name_a',
    desc: 'desc_a',
  }">
    <p :data-bbb="{
      type: 'type_b',
      name: 'name_b',
      desc: 'desc_b',
    }">
    </p>
  </div>`;
  test('when key is regExp', () => {
    const vueKeyREMap = new Map();
    const dataKey = /data-.+/;
    const keys = resolveDataKey(dataKey);
    keys.map((item, index) => vueKeyREMap.set(index, vueKeyRe(`${item}`)));

    expect(transformVueKey(code, vueKeyREMap)?.code).toMatchSnapshot();
  });
});

describe('transformJsxKey', () => {
  const code = `<div
  data-aaa={{
    type: 'type_a',
    name: 'name_a',
    desc: 'desc_a',
  }}
>
  <p
    data-bbb={{
      type: 'type_b',
      name: 'name_b',
      desc: 'desc_b',
    }}
  >
    jsx
  </p>
</div>`;
  test('when key is regExp', () => {
    const jsxKeyREMap = new Map();
    const dataKey = /data-.+/;
    const keys = resolveDataKey(dataKey);
    keys.map((item, index) => jsxKeyREMap.set(index, jsxKeyRe(`${item}`)));

    expect(transformJsxKey(code, jsxKeyREMap)?.code).toMatchSnapshot();
  });
});
