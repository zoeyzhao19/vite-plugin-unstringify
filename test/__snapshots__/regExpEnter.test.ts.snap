// Vitest Snapshot v1

exports[`transformJsxKey > when key is regExp 1`] = `
"<div
  data-aaa={{JSON.stringify({
    type: 'type_a',
    name: 'name_a',
    desc: 'desc_a',
  })
>
  <p
    data-bbb={{JSON.stringify({
      type: 'type_b',
      name: 'name_b',
      desc: 'desc_b',
    })
  >
    jsx
  </p>
</div>"
`;

exports[`transformVueKey > when key is regExp 1`] = `
"<div :data-aaa=\\"JSON.stringify({
    type: 'type_a',
    name: 'name_a',
    desc: 'desc_a',
  })\\">
    <p :data-bbb=\\"JSON.stringify({
      type: 'type_b',
      name: 'name_b',
      desc: 'desc_b',
    })\\">
    </p>
  </div>"
`;
